src/protocols/user_directed_commissioning/UserDirectedCommissioning.h

```diff
@@ -0,0 +1,247 @@
+
+    virtual ~UserConfirmationProvider() {}
+};
+
+class DLL_EXPORT UserDirectedCommissioningClient : public Messaging::ExchangeDelegate
+{
+public:
+    // TODO: Init function will take a Channel instead a SecureSessionHandle, when Channel API is ready
+    /**
+     *  Initialize the UserDirectedCommissioningClient object. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    exchangeMgr    A pointer to the ExchangeManager object.
+     *  @param[in]    sessoin        A handle to the session.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(Messaging::ExchangeManager * exchangeMgr, SecureSessionHandle session);
+
+    /**
+     *  Shutdown the UserDirectedCommissioningClient. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     */
+    void Shutdown();
+
+    /**
+     * Send a User Directed Commissioning message to a CHIP node.
+     *
+     * @param payload       A PacketBufferHandle with the payload.
+     * @param sendFlags     Flags set by the application for the CHIP message being sent.
+     *
+     * @return CHIP_ERROR_NO_MEMORY if no ExchangeContext is available.
+     *         Other CHIP_ERROR codes as returned by the lower layers.
+     *
+     */
+    CHIP_ERROR
+    SendUDCRequest(System::PacketBufferHandle && payload,
+                   const Messaging::SendFlags & sendFlags = Messaging::SendFlags(Messaging::SendMessageFlags::kNoAutoRequestAck));
+
+private:
+    Messaging::ExchangeManager * mExchangeMgr = nullptr;
+    Messaging::ExchangeContext * mExchangeCtx = nullptr;
+    SecureSessionHandle mSecureSession;
+
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PacketHeader & packetHeader,
+                                 const PayloadHeader & payloadHeader, System::PacketBufferHandle && payload) override;
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
+};
+
+class DLL_EXPORT UserDirectedCommissioningServer : public Messaging::ExchangeDelegate
+{
+public:
+    /**
+     *  Initialize the UserDirectedCommissioningServer object. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    exchangeMgr    A pointer to the ExchangeManager object.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(Messaging::ExchangeManager * exchangeMgr);
+
+    static UserDirectedCommissioningServer & GetInstance()
```

Comment:
(Not a request, just a question) - the server has this GetInstance() function, implying it's meant as a singleton. The client doesn't - does that mean there can / should be multiple clients?

Reply:
For the server its useful to have a singleton since different parts of the code want a handle to the global instance. For the client, there is no reason for a singleton since it just sends a message and then can go away.